src/implicit_function.jl

```diff
@@ -39,79 +36,80 @@ end
 """
     implicit(x)
 
-Make [`ImplicitFunction`](@ref) callable by applying `implicit.forward`.
+Make [`ImplicitFunction{F,C,L}`](@ref) callable by applying `implicit.forward`.
 """
 (implicit::ImplicitFunction)(x) = implicit.forward(x)
 
 """
     frule(rc, (_, dx), implicit, x)
 
-Custom forward rule for [`ImplicitFunction`](@ref).
+Custom forward rule for [`ImplicitFunction{F,C,L}`](@ref).
 
 We compute the Jacobian-vector product `Jv` by solving `Au = Bv` and setting `Jv = u`.
 """
 function ChainRulesCore.frule(
-    rc::RuleConfig, (_, dx), implicit::ImplicitFunction, x::AbstractVector
-)
+    rc::RuleConfig, (_, dx), implicit::ImplicitFunction, x::AbstractArray{R}
+) where {R<:Real}
     (; forward, conditions, linear_solver) = implicit
 
     y = forward(x)
-    n, m = length(x), length(y)
 
     conditions_x(x̃) = conditions(x̃, y)
     conditions_y(ỹ) = -conditions(x, ỹ)
 
     pushforward_A(dỹ) = frule_via_ad(rc, (NoTangent(), dỹ), conditions_y, y)[2]
     pushforward_B(dx̃) = frule_via_ad(rc, (NoTangent(), dx̃), conditions_x, x)[2]
 
-    mul_A!(res, v) = res .= pushforward_A(v)
-    mul_B!(res, v) = res .= pushforward_B(v)
+    mul_A!(res, u::AbstractVector) = res .= vec(pushforward_A(reshape(u, size(y))))
+    mul_B!(res, v::AbstractVector) = res .= vec(pushforward_B(reshape(v, size(x))))
 
-    A = LinearOperator(Float64, m, m, false, false, mul_A!)
-    B = LinearOperator(Float64, m, n, false, false, mul_B!)
+    n, m = length(x), length(y)
+    A = LinearOperator(R, m, m, false, false, mul_A!)
+    B = LinearOperator(R, m, n, false, false, mul_B!)
 
-    dx_vec = Vector(unthunk(dx))
+    dx_vec = convert(Vector{R}, vec(unthunk(dx)))
```

Comment:
why is the conversion here necessary?

Reply:
Because the linear solvers from Krylov.jl for `Au = v` try to initialize a solution `u` of the same type as the right-hand `v`. This result in an error when `v` is, for instance, a `FillArray`.
Here I used `convert` instead of `Vector` in the hope that if `dx` is already a `Vector`, no allocation happens (see [here](https://docs.julialang.org/en/v1/manual/conversion-and-promotion/#Mutable-collections)). However, I'm not sure the `Vector{R}` is necessary, `Vector` would probably suffice?

Reply:
I see. Thanks for clarifying.

Reply:
`Vector{R}` is fine I think